plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSink.java

```diff
@@ -521,24 +523,70 @@ public static Object getIcebergValue(Block block, int position, Type type)
         throw new UnsupportedOperationException("Type not supported as partition column: " + type.getDisplayName());
     }
 
-    private static List<PartitionColumn> toPartitionColumns(List<IcebergColumnHandle> handles, PartitionSpec partitionSpec)
+    private static List<PartitionColumn> toPartitionColumns(List<IcebergColumnHandle> handles, PartitionSpec partitionSpec, Schema schema)
     {
         Map<Integer, Integer> idChannels = new HashMap<>();
         for (int i = 0; i < handles.size(); i++) {
             idChannels.put(handles.get(i).getId(), i);
         }
 
         return partitionSpec.fields().stream()
-                .map(field -> {
-                    Integer channel = idChannels.get(field.sourceId());
-                    checkArgument(channel != null, "partition field not found: %s", field);
-                    Type inputType = handles.get(channel).getType();
-                    ColumnTransform transform = getColumnTransform(field, inputType);
-                    return new PartitionColumn(field, channel, inputType, transform.getType(), transform.getBlockTransform());
-                })
+                .map(field -> getPartitionColumn(field, handles, schema.asStruct(), idChannels))
                 .collect(toImmutableList());
     }
 
+    private static PartitionColumn getPartitionColumn(PartitionField field, List<IcebergColumnHandle> handles, Types.StructType schema, Map<Integer, Integer> idChannels)
+    {
+        List<Integer> sourceIds = getIndexPathToField(schema, getNestedFieldNames(schema, field.sourceId()));
```

Comment:
Can't we slightly change the `getIndexPathToField` algorithm to work with `.fieldId()` instead of `.name()` for the NestedField ?
We seem to get the names just to transform them to ids